packages/runtime/src/runtimes/react/controls/rich-text.tsx

```diff
@@ -1,161 +1,26 @@
-
-import { Value } from 'slate'
-import { Editor, OnChangeParam } from 'slate-react'
-import { useIsInBuilder } from '../../../react'
-import { pollBoxModel } from '../pollBoxModel'
-export type RichTextControlValue = ReactNode
-
-const COMMIT_DEBOUNCE_DELAY = 500
-
-export function useRichText(
-  data: RichTextControlData,
-  control: RichTextControl | null,
-  descriptor: RichTextControlDefinition,
-) {
-  const [editor, setEditor] = useState<Editor | null>(null)
-
-  useEffect(() => {
-    const element = editor?.findDOMNode([])
-    if (element == null || control == null) return
-    return pollBoxModel({
-      element: element as unknown as HTMLElement,
-      onBoxModelChange: boxModel => control.changeLayout(boxModel),
-    })
-  }, [editor, control])
 
-  useEffect(() => {
-    if (editor) control?.setSlateEditor(editor)
-  }, [control, editor])
+import { PropControllersHandle } from '../../../state/modules/prop-controller-handles'
+import Text from '../../../components/builtin/Text'
 
-  /**
-   * We must keep local state so that we can reflect the user's typed changes immediately. At the
-   * same time, though, the source of truth for the data is the prop data. This presents a
-   * challenge: how do we keep local state in sync with the prop data without mangling user input as
-   * data comes in?
-   *
-   * Consider, for example, that the user types "Hello". If at a later time, when the user is trying
-   * to type ", world" the component re-renders with prop data "H", "He", "Hel", "Hell", "Hello", it
-   * will disrupt the user's typing. This could also happen as a result of the prop data changing
-   * for other reasons, like collaborators changing the prop data concurrently. We want to avoid to
-   * disrupt the user's typing, while at the same time display the "true" value as quickly as
-   * possible.
-   *
-   * The approach we take here is to commit the prop data at an opportune time: as the user is
-   * typing we avoid to commit prop data. But once they've stopped typing, we commit it as soon as
-   * possible. This is known as a debounce.
-   */
-  const [value, setValue] = useState(() => {
-    const { selection, ...textWithoutSelection } = data ?? descriptor.config.defaultValue
+export type RichTextControlValue = ReactNode
 
-    return Value.fromJSON(textWithoutSelection)
-  })
-  const [shouldCommit, setShouldCommit] = useState(true)
+export type Descriptors = { text?: RichTextControlDefinition }
 
-  useEffect(() => {
-    if (shouldCommit) {
-      const nextValue = Value.fromJSON(data ?? descriptor.config.defaultValue)
-
-      setValue(currentValue =>
-        currentValue.selection.isBlurred
-          ? Value.fromJSON(nextValue.toJSON({ preserveSelection: false }))
-          : nextValue,
-      )
-    }
-  }, [shouldCommit, data])
+export function useRichText(data: RichTextControlData, control: RichTextControl | null) {
+  const [current, ref] = useState<PropControllersHandle<Descriptors> | null>(null)
```

Comment:
Were you expecting me to be able to do this with a `useRef` ?

I don't think it's possible, but I wanted to double check.

Reply:
Yes, we should be able to avoid using state here

Reply:
I don't think we can avoid it actually. 

Tried again and I don't think we can on second thought. 

We want to set the controller in the effect so it has to be in state. Because `useEffect` wont rerun on ref change.

I could put the `set` in a ref callback, but then that is not reactive to my controller changing.

Reply:
I was wrong here.

Using a ref callback with `useCallback` actually nicely keeps this out of state yet reactive to the controller changing.